src/tree/TreeItemState.ts

```diff
@@ -0,0 +1,141 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { createGenericItem } from '../utils/GenericItem';
+import { TreeElementBase } from './ContainerAppsBranchDataProvider';
+
+interface TreeItemState {
+    /**
+     * Apply a temporary description to the tree item
+     */
+    temporaryDescription?: string;
+    /**
+     * Set the tree item icon to a spinner
+     */
+    spinner?: boolean;
+    /**
+     * Temporary children to be displayed
+     */
+    temporaryChildren?: TreeElementBase[];
+}
+
+type ResourceGroupsItem = TreeElementBase & { id: string };
+
+export class TreeItemStateStore implements vscode.Disposable {
```

Comment:
Seems like we should do some code sharing with this class/file.

Reply:
Yes. I will put an implementation in utils. Planning on making it cover our use cases, but allow for clients to extend if needed.